### ISSUE TITLE
pod2html mistranslates ordered lists

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6958]
Migrated from [rt.perl.org#24591](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=24591) (status was 'resolved')

Searchable as RT24591$

**Comment by p5pRT** [comment]
### From kodis@acm.org

#### Created by kodis@acm.org



The pod2html utility misgenerates html for ordered and unordered
lists\.  Here's an example\.  This pod file&#8203;:

=head1 Bugs in pod2html

The problem is that I'd really like ordered and unordered lists\, but
can't get pod2html to generate them for me\.  Watch\.\.\.

This should be an unordered list&#8203;:

=over

=item \*

The first item is item one\,

=item \*

The second item is item two\.\.\.

=back

This should be an ordered list&#8203;:

=over

=item 1\.

The first item is item one\,

=item 2\.

The second item is item two\.\.\.

=back

\.\.\. generates the following html \(I've snipped the boilerplate at the
beginning of the html\)&#8203;:

\<h1>\<a name="bugs\_in\_pod2html">Bugs in pod2html\</a>\</h1>
\<p>The problem is that I'd really like ordered and unordered lists\, but
can't get pod2html to generate them for me\.  Watch\.\.\.\</p>
\<p>This should be an unordered list&#8203;:\</p>
\<ul>
\<li>\</li>
The first item is item one\,
\<p>\</p>
\<li>\</li>
The second item is item two\.\.\.
\<p>\</p>\</ul>
\<p>This should be an ordered list&#8203;:\</p>
\<ol>
\<li>\</li>
The first item is item one\,
\<p>\</p>
\<li>\</li>
The second item is item two\.\.\.
\<p>\</p>\</ol>

In both cases\, the text of the list elements gets placed outside of
the \<li> and \<p> tags\.  I've tried all the variations that I can think
of \(=item\, =item 1\, =item 1\.\, with the item text immediately
following\, seperated by a newline\, seperated by a blank line\, and a
few other less likely versions\)\, but am unable to get any type of
reasonable looking list generated\.

<details><summary>Perl Info</summary>

```

Flags:
    category=utilities
    severity=medium

Site configuration information for perl v5.8.0:

Configured by bhcompile at Wed Aug 13 11:45:59 EDT 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.21-1.1931.2.382.entsmp, archname=i386-linux-thread-multi
    uname='linux stripples.devel.redhat.com 2.4.21-1.1931.2.382.entsmp #1 smp wed aug 6 17:18:52 edt 2003 i686 i686 i386 gnulinux '
    config_args='-des -Doptimize=-O2 -g -pipe -march=i386 -mcpu=i686 -Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red Hat, Inc. -Dinstallprefix=/usr -Dprefix=/usr -Darchname=i386-linux -Dvendorprefix=/usr -Dsiteprefix=/usr -Dotherlibdirs=/usr/lib/perl5/5.8.0 -Duseshrplib -Dusethreads -Duseithreads -Duselargefiles -Dd_dosuid -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2 -g -pipe -march=i386 -mcpu=i686',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='3.2.2 20030222 (Red Hat Linux 3.2.2-5)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.8.0/i386-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    MAINT18379


@INC for perl v5.8.0:
    /usr/lib/perl5/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/5.8.0
    /usr/lib/perl5/site_perl/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/vendor_perl/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.0
    /usr/lib/perl5/vendor_perl
    /usr/lib/perl5/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/5.8.0
    .


Environment for perl v5.8.0:
    HOME=/home/kodis
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/kodis/bin:/home/kodis/sbin:/usr/local/bin:/usr/local/sbin:/usr/kerberos/bin:/usr/kerberos/sbin:/usr/X11R6/bin:/usr/bin:/usr/sbin:/bin:/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From sburke@cpan.org

At 09&#8203;:02 AM 2003\-12\-03\, Kodis@&#8203;Acm\.Org wrote&#8203;:
>The pod2html utility misgenerates html for ordered and unordered
>lists\.  Here's an example\.  This pod file&#8203;:

Yes\, pod2html is crap\.  We're totally replacing it soon\, don't worry\.

\-\-
Sean M\. Burke    http&#8203;://search\.cpan\.org/~sburke/

**Comment by p5pRT** [comment]
### From @schwern

On Wed\, Dec 03\, 2003 at 06&#8203;:02&#8203;:26PM \-0000\, kodis@&#8203;acm\.org \(via RT\) wrote&#8203;:
> In both cases\, the text of the list elements gets placed outside of
> the \<li> and \<p> tags\.  I've tried all the variations that I can think
> of \(=item\, =item 1\, =item 1\.\, with the item text immediately
> following\, seperated by a newline\, seperated by a blank line\, and a
> few other less likely versions\)\, but am unable to get any type of
> reasonable looking list generated\.

Pod&#8203;::Html wants you to do this\.

=over 4

=item \* The first item is item one\,

=item \* The second item is item two\.\.\.

=back

To get this\.

\<ul>
\<li>The first item is item one\,\</li>
\<li>The second item is item two\.\.\.\</li>
\</ul>

Which is ok\.  But it flubbs your example\.  A very simple way to fix this is
to just not use \</li>\. 

\-\-\- lib/Pod/Html\.pm    2003/12/04 20&#8203;:34&#8203;:42    1\.1
\+\+\+ lib/Pod/Html\.pm    2003/12/04 20&#8203;:39&#8203;:31
@&#8203;@&#8203; \-1152\,7 \+1152\,6 @&#8203;@&#8203;
&nbsp;         \}
&nbsp;         $need\_dd = 1;
&nbsp;     \}
\-    print HTML "\</$emitted>" if $emitted;
&nbsp;     print HTML "\\n";
&nbsp;     return $need\_dd;
 \}

This makes all three styles render something like correctly\.

=over

=item \*

foo

=item \* foo

=item \* foo

bar

=back


\-\- 
Michael G Schwern        schwern@&#8203;pobox\.com  http&#8203;://www\.pobox\.com/~schwern/
We don't know\.  But if we did\, we wouldn't tell you\.

**Comment by p5pRT** [comment]
### From kodis@mail630.gsfc.nasa.gov

On Thu\, Dec 04\, 2003 at 08&#8203;:42&#8203;:30PM \-0000\, Michael G Schwern via RT wrote&#8203;:

> Pod&#8203;::Html wants you to do this\.
> 
> =over 4
> 
> =item \* The first item is item one\,
> 
> =item \* The second item is item two\.\.\.
> 
> =back
> 
> To get this\.
> 
> \<ul>
> \<li>The first item is item one\,\</li>
> \<li>The second item is item two\.\.\.\</li>
> \</ul>

Ah\, yes\.  I tried that\, and while it does put the items within the
list element tags\, it includes a superfluous set of \<strong> tags\.


> \-    print HTML "\</$emitted>" if $emitted;

Hey\, that looks like the way to go\.  Let me play with it on some
samples and see how it works\.

Thanks for looking into the problem\.

\-\- 
John Kodis                                    Goddard Space Flight Center
kodis@&#8203;mail630\.gsfc\.nasa\.gov                      Greenbelt\, Maryland\, USA
Phone&#8203;: 301\-286\-7376                                     Fax&#8203;: 301\-286\-1771

**Comment by p5pRT** [comment]
### From nick.ing-simmons@elixent.com

Sean M\. Burke \<sburke@&#8203;cpan\.org> writes&#8203;:
>At 09&#8203;:02 AM 2003\-12\-03\, Kodis@&#8203;Acm\.Org wrote&#8203;:
>>The pod2html utility misgenerates html for ordered and unordered
>>lists\.  Here's an example\.  This pod file&#8203;:
>
>Yes\, pod2html is crap\.  We're totally replacing it soon\, don't worry\.

The beginings at least of one that uses pod parser of its vintage 
is in CPAN/authors/id/N/NI/NI\-S/PodToHTML\-0\.04\.tar\.gz

**Comment by p5pRT** [comment]
### From dland@landgren.net

On Wed Dec 03 10&#8203;:02&#8203;:25 2003\, kodis@&#8203;acm\.org wrote&#8203;:
> 
> This is a bug report for perl from kodis@&#8203;acm\.org\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> The pod2html utility misgenerates html for ordered and unordered
> lists\.  Here's an example\.  This pod file&#8203;:
> 
> =head1 Bugs in pod2html
> 
> The problem is that I'd really like ordered and unordered lists\, but
> can't get pod2html to generate them for me\.  Watch\.\.\.
> 
> This should be an unordered list&#8203;:
> 
> =over
> 
> =item \*
> 
> The first item is item one\,
> 
> =item \*
> 
> The second item is item two\.\.\.
> 
> =back
> 
> This should be an ordered list&#8203;:
> 
> =over
> 
> =item 1\.
> 
> The first item is item one\,
> 
> =item 2\.
> 
> The second item is item two\.\.\.
> 
> =back
> 
> \.\.\. generates the following html \(I've snipped the boilerplate at the
> beginning of the html\)&#8203;:
> 
> \<h1>\<a name="bugs\_in\_pod2html">Bugs in pod2html\</a>\</h1>
> \<p>The problem is that I'd really like ordered and unordered lists\,
>    but
> can't get pod2html to generate them for me\.  Watch\.\.\.\</p>
> \<p>This should be an unordered list&#8203;:\</p>
> \<ul>
> \<li>\</li>
> The first item is item one\,
> \<p>\</p>
> \<li>\</li>
> The second item is item two\.\.\.
> \<p>\</p>\</ul>
> \<p>This should be an ordered list&#8203;:\</p>
> \<ol>
> \<li>\</li>
> The first item is item one\,
> \<p>\</p>
> \<li>\</li>
> The second item is item two\.\.\.
> \<p>\</p>\</ol>

This has been corrected in Pod&#8203;::Html version 1\.09\, which was released
with perl 5\.10\.0\. I am currently working on publishing this module on
CPAN \(as a dual\-life module\)\, hence older perls will be able to benefit
from bugfixes such as this one\.

Thanks\,
David Landgren \(on behalf of the perl\.org team\)

**Comment by p5pRT** [comment]
dland@landgren.net - Status changed from 'open' to 'resolved'
